thermal-service/src/context.rs

```diff
@@ -0,0 +1,140 @@
+    }
+
+    /// Wait for a request on the IPC channel
+    pub(crate) async fn wait_request(&self) -> ipc::Request<NoopRawMutex, Request, mptf::Response> {
+        CONTEXT.get().await.ipc.receive().await
+    }
+
+    /// Wait for a request via the comms service
+    pub(crate) async fn wait_comms_request(&self) -> Request {
+        CONTEXT.get().await.comms.receive().await
+    }
+
+    /// Transfer a request from the comms service
+    pub(crate) fn send_request_from_comms(&self, msg: Request) -> Result<(), ()> {
+        CONTEXT.try_get().ok_or(())?.comms.try_send(msg).map_err(|_| ())?;
+        Ok(())
+    }
+}
+
+// Just one instance of this context should be instantiated
+static CONTEXT: OnceLock<Context> = OnceLock::new();
+
+/// Send a request to the thermal service directly instead of through comms.
+pub async fn execute_service_request(request: Request) -> mptf::Response {
+    CONTEXT.get().await.ipc.execute(request).await
+}
+
+/// Register a sensor with the thermal service
+pub async fn register_sensor(sensor: &'static sensor::Device) -> Result<(), intrusive_list::Error> {
+    if get_sensor(sensor.id()).await.is_some() {
+        return Err(intrusive_list::Error::NodeAlreadyInList);
+    }
+
+    CONTEXT.get().await.sensors.push(sensor)
+}
+
+/// Provides access to the sensors list
+pub async fn sensors() -> &'static intrusive_list::IntrusiveList {
+    &CONTEXT.get().await.sensors
+}
+
+/// Find a sensor by its ID
+pub async fn get_sensor(id: sensor::DeviceId) -> Option<&'static sensor::Device> {
+    for sensor in &CONTEXT.get().await.sensors {
+        if let Some(data) = sensor.data::<sensor::Device>() {
+            if data.id() == id {
+                return Some(data);
+            }
+        } else {
+            error!("Non-device located in sensors list");
+        }
+    }
+
+    None
+}
+
+/// Send a request to a sensor through the thermal service instead of directly.
+pub async fn execute_sensor_request(id: sensor::DeviceId, request: sensor::Request) -> sensor::Response {
+    let sensor = get_sensor(id).await.ok_or(sensor::Error::InvalidRequest)?;
+    sensor.execute_request(request).await
+}
+
+/// Register a fan with the thermal service
+pub async fn register_fan(fan: &'static fan::Device) -> Result<(), intrusive_list::Error> {
+    if get_fan(fan.id()).await.is_some() {
+        return Err(intrusive_list::Error::NodeAlreadyInList);
+    }
+
+    CONTEXT.get().await.fans.push(fan)
+}
+
+/// Provides access to the fans list
+pub async fn fans() -> &'static intrusive_list::IntrusiveList {
+    &CONTEXT.get().await.fans
+}
+
+/// Find a fan by its ID
+pub async fn get_fan(id: fan::DeviceId) -> Option<&'static fan::Device> {
+    for fan in &CONTEXT.get().await.fans {
+        if let Some(data) = fan.data::<fan::Device>() {
+            if data.id() == id {
+                return Some(data);
+            }
+        } else {
+            error!("Non-device located in fan list");
+        }
+    }
+
+    None
```

Comment:
Very similar to the previous function, could be solved similarly:

```suggestion
    fans()
        .await
        .into_iter()
        .filter_map(|sensor| sensor.data::<fan::Device>())
        .find(|data| data.id() == id)
```